src/outer_tracing.jl

```diff
@@ -0,0 +1,257 @@
+        return 1
+    elseif next == Tuple(previous) .+ (-1, 0)
+        return 2
+    elseif next == Tuple(previous) .+ (-1, -1)
+        return 3
+    elseif next == Tuple(previous) .+ (0, -1)
+        return 4
+    elseif next == Tuple(previous) .+ (1, -1)
+        return 5
+    elseif next == Tuple(previous) .+ (1, 0)
+        return  6
+    elseif next == Tuple(previous) .+ (1, 1)
+        return 7
+    end
+end
+
+function invert(dₚc)
+    if dₚc >= 4
+        return dₚc - 4
+    else
+        return dₚc + 4
+    end
+end
+
+function find_next(pc, dₚc, pₙ, dcₙ, pad_box)
+    dcₚ = invert(dₚc)
+    for r in 0:6
+        dₑ = mod(dcₚ + r, 8)
+        dᵢ = mod(dcₚ + r + 1, 8)
+        pₑ = chainpoint(pc, dₑ)
+        pᵢ = chainpoint(pc, dᵢ)
+        if is_pixel_background(pad_box, pₑ) && !is_pixel_background(pad_box, pᵢ)
+            pₙ = pₑ
+            dcₙ = dₑ
+        end
+    end
+    return pₙ, dcₙ
+end
+
+# Translates the mod calculation to coordinates.
+function chainpoint(p, d)
+    if d == 0
+        return p .+ (0, 1)
+    elseif d == 1
+        return p .+ (-1, 1)
+    elseif d == 2
+        return p .+ (-1, 0)
+    elseif d == 3
+        return p .+ (-1, -1)
+    elseif d == 4
+        return p .+ (0, -1)
+    elseif d == 5
+        return p .+ (1, -1)
+    elseif d == 6
+        return p .+ (1, 0)
+    elseif d == 7
+        return p .+ (1, 1)
+    end
+end
+
+# Creates the bounds around each component.
+function bounding_box(img, components)
+    for r in 1:size(img, 1)
+        for c in 1:size(img, 2)
+            if img[r, c] != 0 && components[img[r, c], 1] != 0
+                components[img[r, c], 1] = compare_left(components[img[r, c], 1], c)
+                components[img[r, c], 2] = compare_top(components[img[r, c], 2], r)
+                components[img[r, c], 3] = compare_bottom(components[img[r, c], 3], r)
+                components[img[r, c], 4] = compare_right(components[img[r, c], 4], c)
+            elseif img[r, c] != 0 && components[img[r, c], 1] == 0
+                components[img[r, c], 1] = c
+                components[img[r, c], 2] = r
+                components[img[r, c], 3] = r
+                components[img[r, c], 4] = c
+            end
+        end
+    end
+    return components
+end
+
+function is_pixel_background(pad_box, current_pixel)
+    if (pad_box[CartesianIndex(current_pixel)] == 0)
+        return true
+    else
+        return false
+    end
+end
+
+# Removes the padding from the bounding boxes and restores the view to the
+# original array.
+function restore(boundary, boundary_lengths, components)
+    start = 1
+    for i in 1:length(boundary_lengths)
+        for j in start:start+boundary_lengths[i]-1
+            boundary[j] = boundary[j] .+ (components[i, 2] - 3, components[i, 1] - 3)
+        end
+        start += boundary_lengths[i]
+    end
+    return boundary
+end
+
+# The compare functions are to expand the bounding box only where necessary.
```

Comment:
I noticed that you use the functions in your moore tracing algorithm too, maybe put them in a separate file, so that you do not have to repeat them twice? its not a big deal tho